### PR TITLE
FMC reset flows layout.

### PR DESCRIPTION
**Pull request by rusty1968**
Execute specialized FMC flows based on reset reason.

### CODE REVIEW

File: fmc/src/flow/cold_reset.rs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+/// Warm Reset Flow
+pub struct ColdResetFlow {}
+
+impl ColdResetFlow {
+    /// Execute update reset flow
+    ///
+    /// # Arguments
+    ///
+    /// * `env` - FMC Environment
+    #[inline(never)]
```
Comment (jhand2) [inline fmc/src/flow/cold_reset.rs:28]: just curious, why don't we want to inline this?

File: fmc/src/flow/crypto.rs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+Abstract:
+
+    Crypto helper routines
+
+--*/
+use caliptra_drivers::{Ecc384PubKey, KeyId};
+/// DICE  Layer Key Pair
+#[derive(Debug)]
+pub struct Ecc384KeyPair {
```
Comment (jhand2) [inline fmc/src/flow/crypto.rs:17]: Should probably move this into the driver crate and then just use it in both ROM and FMC

**Review by jhand2 — COMMENTED** [review]
Made it about half way. I'll try to get to the rest by end-of-day.

Sorry for the delay in looking at this.

File: fmc/src/flow/crypto.rs (hunk 2 above)
Comment (rusty1968) [reply fmc/src/flow/crypto.rs:17]: I agree, but  Vishal is  trying to minimize code sharing and driver crate is supposed to be ROM only. I can move the type to the common crate.

Comment (rusty1968) [reply fmc/src/flow/crypto.rs]: Done

File: fmc/src/flow/dice.rs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    pub subj_sn: [u8; 64],
+
+    /// Subject Key Identifier
+    pub subj_key_id: [u8; 20],
+
+    /// Runtime Load Address
+    pub runtime_load_addr: u32,
+
+    /// Runtime Entry Point
+    pub runtime_entry_point: u32,
```
[the remark this answers is not in the document]
Comment (rusty1968) [reply fmc/src/flow/dice.rs]: I borrowed these from ROM but they are unused so just removed.

File: fmc/src/flow/cold_reset.rs (hunk 1 above)
Comment (rusty1968) [reply fmc/src/flow/cold_reset.rs:28]: Make it easy for debug.

File: fmc/src/flow/dice.rs [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    ///
+    /// Layer Private Key will be generated in the Key Vault slot specified
+    /// by this slot
+    pub subj_priv_key: KeyId,
+
+    /// Temporary KeyId used during DICE derivations
+    pub uds_key: KeyId,
+
+    /// Temporary KeyId used during DICE derivations
+    pub fe_key: KeyId,
```
[the remark this answers is not in the document]
Comment (rusty1968) [reply fmc/src/flow/dice.rs:50]: Field entropy. Added comment.

File: fmc/src/flow/mod.rs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+///
+/// # Arguments
+///
+/// * `env` - FMC Environment
+pub fn run(env: &FmcEnv, hand_off: &mut HandOff) -> CaliptraResult<()> {
+    // Placeholder.
+    if let Some(input) = try_from_hand_off(env, hand_off) {
+        let _ = RtAliasLayer::derive(env, &input);
+    } else {
+        // TODO : handle error
```
Comment (jhand2) [inline fmc/src/flow/mod.rs:46]: Should this error just be passed up the call stack and dealt with in main?

File: fmc/src/main.rs [hunk 6]
```diff
@@ -35,12 +36,13 @@ Running Caliptra FMC ...
...
 
-    if let Some(hand_off) = HandOff::from_previous() {
+    if let Some(mut hand_off) = HandOff::from_previous() {
         let env = fmc_env::FmcEnv::default();
-        hand_off.to_rt(&env)
-    } else {
-        caliptra_drivers::ExitCtrl::exit(0xff)
+        if flow::run(&env, &mut hand_off).is_ok() {
+            hand_off.to_rt(&env)
+        }
```
Comment (jhand2) [inline fmc/src/main.rs:43]: Don't need to decide in this PR, but if something fails during boot we probably want to do one of

1. Try again to boot from rom (maybe invoke an impactless reset?)
2. Notify the SoC and then wait for a new firmware image

This is probably something to discuss with the group.

Comment (rusty1968) [reply fmc/src/main.rs:43]: I will file an issue.

File: fmc/src/flow/mod.rs (hunk 5 above)
Comment (rusty1968) [reply fmc/src/flow/mod.rs:46]: I am making flow our top level module for FMC logic and entry_point  (main) just a Rust entry point, but I am open to other suggestions :-)

Comment (jhand2) [reply fmc/src/flow/mod.rs:46]: works for me

**Review by jhand2 — APPROVED** [review]

**Comment by rusty1968** [comment]
@jhand2  do you have write permissions to merge?

**Comment by jhand2** [comment]
> @jhand2 do you have write permissions to merge?

I do not :)

cc @korran

File: fmc/src/flow/dice.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (rusty1968) [reply fmc/src/flow/dice.rs]: I removed them.

**Pull request merged**